front/lib/api/files/client_executable.ts

```diff
@@ -277,3 +291,398 @@ export async function getClientExecutableFileContent(
+    )
+    .map(([, actions]) => actions);
+
+  // Remaining edit-only groups to cancel. Starts from the current (external) revert.
+  let cancelGroupActionCounter = revertCount;
+
+  const pickedEditActions = [];
+
+  for (const actionGroup of sortedActionGroups) {
+    if (cancelGroupActionCounter > 0) {
+      const revertActions = actionGroup.filter((a) =>
+        isRevertFileActionType(a)
+      );
+
+      // Cancel everything if there are only edit actions in the group.
+      if (revertActions.length === 0) {
+        cancelGroupActionCounter--;
+        continue;
+      }
+
+      // Extend cancellation window by each revert's count (any edits will be cancelled).
+      for (const revertAction of revertActions) {
+        const count =
+          typeof revertAction.augmentedInputs.revertCount === "number"
+            ? revertAction.augmentedInputs.revertCount
+            : 1;
+        cancelGroupActionCounter += count;
+      }
+
+      continue;
+    }
+
+    // Not cancelling: collect edits. A revert here reopens cancellation for older groups.
+    for (const currentAction of actionGroup) {
+      if (isEditFileActionType(currentAction)) {
+        pickedEditActions.push(currentAction);
+        continue;
+      }
+
+      if (isRevertFileActionType(currentAction)) {
+        const count =
+          typeof currentAction.augmentedInputs.revertCount === "number"
+            ? currentAction.augmentedInputs.revertCount
+            : 1;
+        cancelGroupActionCounter += count;
+      }
+    }
+  }
+
+  return pickedEditActions;
+}
+
+export function getRevertedContent(
+  createFileAction: AgentMCPActionModel,
+  actionsToApply: AgentMCPActionModel[]
+) {
+  // Sort actions from oldest to latest since we need to apply the changes from top to bottom.
+  const sortedActions = actionsToApply.sort(
+    (a, b) => a.createdAt.getTime() - b.createdAt.getTime()
+  );
+
+  if (!isCreateFileActionType(createFileAction)) {
+    throw new Error("Wrong file action type for create file action");
+  }
+
+  let revertedContent = createFileAction.augmentedInputs.content;
+
+  for (const editAction of sortedActions) {
+    if (!isEditFileActionType(editAction)) {
+      throw new Error("Wrong file action type for edit file action");
+    }
+
+    const { old_string, new_string } = editAction.augmentedInputs;
+
+    const { updatedContent, occurrences } = getUpdatedContentAndOccurrences(
+      old_string,
+      new_string,
+      revertedContent
+    );
+
+    if (occurrences === 0) {
+      throw new Error(`Cannot find matched text: "${old_string}"`);
+    }
+
+    revertedContent = updatedContent;
+  }
+
+  return revertedContent;
+}
+
+// Revert the changes made to the Content Creation file in the last agent message.
+// This reconstructs the previous file state by replaying edit operations chronologically
+// from the create action.
+export async function revertClientExecutableFileChanges(
```

Comment:
Any testable part in this logic I extracted to a function, this file is a bit too big now and I can move those functions to utils, but not sure if it's easier to read if I move them all 🤔